File: hard/0480_sliding_window_median/solution.cpp
#include <functional>
#include <queue>
#include <vector>

class Solution {
public:
  std::vector<double> medianSlidingWindow(std::vector<int> &nums, int k) {
    std::priority_queue<std::pair<int, size_t>> maxHeap;
    std::priority_queue<std::pair<int, size_t>,
                        std::vector<std::pair<int, size_t>>,
                        std::greater<std::pair<int, size_t>>>
        minHeap;

    std::vector<int> heapMap(nums.size(), -1);
    std::vector<int> staleMap(nums.size(), 0);

    size_t i{};
    for (; i < k; ++i) {
      minHeap.push({nums[i], i});
      heapMap[i] = 1;
    }

    for (size_t _{}; _ < k / 2; ++_) {
      auto [num, j] = minHeap.top();
      minHeap.pop();
      maxHeap.push({num, j});
      heapMap[j] = 0;
    }

    size_t maxHeapSize = maxHeap.size();
    size_t minHeapSize = minHeap.size();

    std::vector<double> medians;
    medians.reserve(nums.size() - k);

    if (k % 2 == 1)
      medians.push_back(minHeap.top().first);
    else
      medians.push_back(maxHeap.top().first / 2.0 + minHeap.top().first / 2.0);

    for (; i < nums.size(); ++i) {
      staleMap[i - k] = staleMap[i - k] + 1;

      if (heapMap[i - k] == 0)
        --maxHeapSize;
      else
        --minHeapSize;

      if (!maxHeap.empty() && nums[i] <= maxHeap.top().first) {
        maxHeap.push({nums[i], i});
        heapMap[i] = 0;
        ++maxHeapSize;
      } else {
        minHeap.push({nums[i], i});
        heapMap[i] = 1;
        ++minHeapSize;
      }

      if (minHeapSize > maxHeapSize + 1) {
        auto [num, j] = minHeap.top();
        minHeap.pop();

        while (staleMap[j]) {
          --staleMap[j];
          std::tie(num, j) = minHeap.top();
          minHeap.pop();
        }

        maxHeap.push({num, j});

        heapMap[j] = 0;
        ++maxHeapSize;
        --minHeapSize;
      } else if (maxHeapSize > minHeapSize) {
        auto [num, j] = maxHeap.top();
        maxHeap.pop();

        while (staleMap[j]) {
          --staleMap[j];
          std::tie(num, j) = maxHeap.top();
          maxHeap.pop();
        }

        minHeap.push({num, j});

        heapMap[j] = 1;
        --maxHeapSize;
        ++minHeapSize;
      }

      while (!maxHeap.empty() && staleMap[maxHeap.top().second]) {
        --staleMap[maxHeap.top().second];
        maxHeap.pop();
      }
      while (!minHeap.empty() && staleMap[minHeap.top().second]) {
        --staleMap[minHeap.top().second];
        minHeap.pop();
      }

      if ((maxHeapSize + minHeapSize) % 2 == 1)
        medians.push_back(minHeap.top().first);
      else
        medians.push_back(maxHeap.top().first / 2.0 +
                          minHeap.top().first / 2.0);
    }

    return medians;
  }
};

File: hard/0480_sliding_window_median/solution.py
import heapq
from typing import List


class Solution:
    def medianSlidingWindow(self, nums: List[int], k: int) -> List[float]:
        maxHeap = []
        minHeap = [(nums[0], 0)]

        heapMap = [-1] * len(nums)
        heapMap[0] = 0

        for i in range(1, k):
            if nums[i] < minHeap[0][0]:
                heapq.heappush(maxHeap, (-nums[i], i))
                heapMap[i] = 1
            else:
                heapq.heappush(minHeap, (nums[i], i))
                heapMap[i] = 0

            if len(maxHeap) < len(minHeap) - 1:
                num, j = heapq.heappop(minHeap)
                heapq.heappush(maxHeap, (-num, j))
                heapMap[j] = 1
            elif len(minHeap) < len(maxHeap):
                num, j = heapq.heappop(maxHeap)
                heapq.heappush(minHeap, (-num, j))
                heapMap[j] = 0

        medians = []

        if k % 2 == 1:
            medians.append(minHeap[0][0])
        else:
            medians.append((-maxHeap[0][0] + minHeap[0][0]) / 2)

        staleMap = [0] * len(nums)

        maxHeapSize = len(maxHeap)
        minHeapSize = len(minHeap)

        for i in range(k, len(nums)):
            staleMap[i - k] = staleMap[i - k] + 1

            if heapMap[i - k] == 1:
                maxHeapSize -= 1
            else:
                minHeapSize -= 1

            if maxHeap and nums[i] <= -maxHeap[0][0]:
                heapq.heappush(maxHeap, (-nums[i], i))
                heapMap[i] = 1
                maxHeapSize += 1
            else:
                heapq.heappush(minHeap, (nums[i], i))
                heapMap[i] = 0
                minHeapSize += 1

            if minHeapSize > maxHeapSize + 1:
                num, j = heapq.heappop(minHeap)

                while staleMap[j]:
                    staleMap[j] -= 1
                    num, j = heapq.heappop(minHeap)

                heapq.heappush(maxHeap, (-num, j))

                heapMap[j] = 1
                maxHeapSize += 1
                minHeapSize -= 1
            elif maxHeapSize > minHeapSize:
                num, j = heapq.heappop(maxHeap)

                while staleMap[j]:
                    staleMap[j] -= 1
                    num, j = heapq.heappop(maxHeap)

                heapq.heappush(minHeap, (-num, j))

                heapMap[j] = 0
                minHeapSize += 1
                maxHeapSize -= 1

            while maxHeap and staleMap[maxHeap[0][1]]:
                staleMap[maxHeap[0][1]] -= 1
                heapq.heappop(maxHeap)
            while minHeap and staleMap[minHeap[0][1]]:
                staleMap[minHeap[0][1]] -= 1
                heapq.heappop(minHeap)

            if (maxHeapSize + minHeapSize) % 2 == 1:
                medians.append(minHeap[0][0])
            else:
                medians.append((-maxHeap[0][0] + minHeap[0][0]) / 2)

        return medians
